File: packages/expo-av/android/src/main/java/expo/modules/av/player/PlayerControl.java
package expo.modules.av.player;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.widget.MediaController;

public class PlayerControl implements MediaController.MediaPlayerControl {
  private final PlayerManager mPlayerManager;

  public PlayerControl(final @NonNull PlayerManager playerManager) {
    mPlayerManager = playerManager;
  }

  @Override
  public void start() {
    final Bundle map = new Bundle();
    map.putBoolean(PlayerManager.STATUS_SHOULD_PLAY_KEY_PATH, true);
    mPlayerManager.setStatus(map, null);
  }

  @Override
  public void pause() {
    final Bundle map = new Bundle();
    map.putBoolean(PlayerManager.STATUS_SHOULD_PLAY_KEY_PATH, false);
    mPlayerManager.setStatus(map, null);
  }

  @Override
  public int getDuration() {
    final Bundle status = mPlayerManager.getStatus();
    return (status.getBoolean(PlayerManager.STATUS_IS_LOADED_KEY_PATH)
        && status.containsKey(PlayerManager.STATUS_DURATION_MILLIS_KEY_PATH))
        ? status.getInt(PlayerManager.STATUS_DURATION_MILLIS_KEY_PATH) : 0;
  }

  @Override
  public int getCurrentPosition() {
    final Bundle status = mPlayerManager.getStatus();
    return status.getBoolean(PlayerManager.STATUS_IS_LOADED_KEY_PATH)
        ? status.getInt(PlayerManager.STATUS_POSITION_MILLIS_KEY_PATH) : 0;
  }

  @Override
  public void seekTo(final int msec) {
    final Bundle map = new Bundle();
    map.putDouble(PlayerManager.STATUS_POSITION_MILLIS_KEY_PATH, (double) msec);
    mPlayerManager.setStatus(map, null);
  }

  @Override
  public boolean isPlaying() {
    final Bundle status = mPlayerManager.getStatus();
    return status.getBoolean(PlayerManager.STATUS_IS_LOADED_KEY_PATH)
        && status.getBoolean(PlayerManager.STATUS_IS_PLAYING_KEY_PATH);
  }

  @Override
  public int getBufferPercentage() {
    final Bundle status = mPlayerManager.getStatus();
    if (status.getBoolean(PlayerManager.STATUS_IS_LOADED_KEY_PATH)
        && status.containsKey(PlayerManager.STATUS_DURATION_MILLIS_KEY_PATH)
        && status.containsKey(PlayerManager.STATUS_PLAYABLE_DURATION_MILLIS_KEY_PATH)) {
      final double duration = status.getInt(PlayerManager.STATUS_DURATION_MILLIS_KEY_PATH);
      final double playableDuration = status.getInt(PlayerManager.STATUS_PLAYABLE_DURATION_MILLIS_KEY_PATH);
      return (int) (playableDuration / duration * 100.0);
    } else {
      return 0;
    }
  }

  @Override
  public boolean canPause() {
    return true;
  }

  @Override
  public boolean canSeekBackward() {
    return true;
  }

  @Override
  public boolean canSeekForward() {
    return true;
  }

  @Override
  public int getAudioSessionId() {
    return mPlayerManager.getAudioSessionId();
  }

  public boolean isFullscreen() {
    return mPlayerManager.isPresentedFullscreen();
  }

  public void toggleFullscreen() {
    mPlayerManager.toggleFullscreen();
  }
}

File: packages/expo-av/android/src/main/java/expo/modules/av/player/PlayerControl.kt
package expo.modules.av.player

import android.widget.MediaController

class PlayerControl(private val mPlayerManager: PlayerManager) :
    MediaController.MediaPlayerControl {

  val isFullscreen: Boolean
    get() = mPlayerManager.isPresentedFullscreen

  override fun start() {
    mPlayerManager.play()
  }

  override fun pause() {
    mPlayerManager.pause()
  }

  override fun seekTo(positionMillis: Int) {
    mPlayerManager.seekTo(positionMillis)
  }

  override fun getDuration(): Int {
    return mPlayerManager.status.durationInMillis
  }

  override fun getCurrentPosition(): Int {
    return mPlayerManager.status.positionMillis
  }

  override fun isPlaying(): Boolean {
    return mPlayerManager.isPlaying
  }

  override fun getBufferPercentage(): Int {
    val status = mPlayerManager.status
    return if (status.isLoaded && status.durationInMillis > 0 && status.playableDurationMillis > 0) {
      (status.playableDurationMillis / status.durationInMillis * 100.0).toInt()
    } else {
      0
    }
  }

  override fun canPause(): Boolean {
    return true
  }

  override fun canSeekBackward(): Boolean {
    return true
  }

  override fun canSeekForward(): Boolean {
    return true
  }

  override fun getAudioSessionId(): Int {
    return mPlayerManager.audioSessionId
  }

  fun toggleFullscreen() {
    mPlayerManager.toggleFullscreen()
  }
}
